doc/firmware_design_guidelines.md

```diff
@@ -0,0 +1,160 @@
+
+```c
+/* Global counter for system-wide use */
+uint16_t system_counter = 0;
+
+/* Function with a uniquely named local variable */
+void sensor_task(void) {
+    uint16_t sensor_task_counter = 0; /* Unique name to avoid shadowing */
+    sensor_task_counter++;
+}
+
+/* AVOID: Shadowing the global variable */
+void incorrect_example(void) {
+    uint16_t system_counter = 0; /* Shadows global system_counter */
+    system_counter++; /* Modifies local variable, not global */
+}
+```
+
+## Integer Types
+
+To ensure clarity, avoid using the `int` type as its size varies across platforms. Instead use fixed-size integer types from `<stdint.h>` (eg. `uint16_t`) to explicitly define the size of integer variables.
+
+Always select a type large enough to prevent integer overflows, which can cause undefined behavior in safety-critical applications.
+
+```c
+/* Explicitly sized variables */
+uint32_t sensor_count = 0; /* 32-bit unsigned for general use */
+uint8_t error_code = 0;    /* 8-bit unsigned for limited range (0-255) */
+
+/* AVOID: Ambiguous size */
+int count = 0; /* Size depends on platform */
+```
+
+## Comments
+
+Every source file must begin with a comment block with our copyright, a high level description of the file's purpose, functionality and its role. If the file controls a specific component or peripheral, include the part number and a link to the component's datasheet for reference. It's recommended to upload the datasheet to Notion to avoid dead links.
+
+```c
+/*
+ * Copyright (c) 2025 Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * EPS Distribution Library
+ *
+ * Firmware for controlling power distribution on the EPS Distribution board.
+ * Functionality includes ...
+ */
+ 
+EPSDistrStatus eps_distribution_init(void);
+```
+
+Functions should contain a comment directly above the function name with a short description of what the function does. It should also separately describe each parameter and the return value if applicable, especially when the parameter or return value is not obvious. Any constraints or acceptable values for the parameters should be specified in the comment.
+
+```c
+/*
+ * Takes in the image dimension N, and uses the image API to return the
+ * prediction of dimension N.
+ *
+ * Returns FINCH_IMG_PREDICT_OK on success and FINCH_IMG_PREDICT_FAIL on error.
+ */
+PredictImageStatus predict_image(
+	const vec3 * N,
+	int32_t prediction[N->z][N->y][N->x]
+);
+```
+
+## File and Folder Organization
+
+Our code is structred into the main app and libraries. The main app contains the `main` function for startup and all other code is put into the `lib` directory. Suppose you are writing the ADCS library. The structure of your code would be as follows,
+
+```
+lib/
+  adcs/
+    *.c
+include/
+  adcs/
+    *.h
+tests/
+  adcs/
+    *.c
+```
```

Comment:
I think adcs/ would be under drivers/ since it interacts with hardware peripherals.

Reply:
also I think "structred" is a typo

Reply:
Ohh I didn't think of that. I thought of a "drivers" folder as like Zephyr drivers but I can make that change